File: widgetcomponent/src/main/java/com/lxch/widgetcomponent/JSWebView.java
package com.lxch.widgetcomponent;

import android.annotation.SuppressLint;
import android.content.Context;
import android.util.AttributeSet;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

/**
 * Created by luoxiaocheng on 2018/3/28.
 */
@SuppressLint("JavascriptInterface")
public class JSWebView extends WebView {
    public String js = "javascript:window.NativeBridge = function(name,message){eval(\"messageHandlers.\"+name+\"(message)\")}";

    public JSWebView(Context context) {
        this(context, null);
    }

    public JSWebView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public JSWebView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        WebSettings settings = getSettings();
        //支持JS
        settings.setJavaScriptEnabled(true);
        //没有缓存
        settings.setCacheMode(WebSettings.LOAD_NO_CACHE);
        //网页自适应屏幕大小
        settings.setLoadWithOverviewMode(true);
        //设置图片适合WebView大小
        settings.setUseWideViewPort(true);
        //图片太大超出屏幕
        settings.setLayoutAlgorithm(WebSettings.LayoutAlgorithm.SINGLE_COLUMN);
        //注入JS代码
        addJavascriptInterface(this, "messageHandlers");
        //添加监听，如果不添加会出现一个空白页面
        setWebChromeClient(new WebChromeClient());
        setWebViewClient(new WebViewClient() {
//            @Override
//            public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
//                return true;
//            }

            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
                //注入JS代码
                view.loadUrl(js);
            }
        });
    }
}

File: widgetcomponent_kotlin/src/main/java/com/lxch/widgetcomponent_kotlin/JSWebView.kt
package com.lxch.widgetcomponent_kotlin

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.webkit.WebChromeClient
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient

/**
 * Created by luoxiaocheng on 2018/3/28.
 */
@SuppressLint("JavascriptInterface")
class JSWebView(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : WebView(context, attrs, defStyleAttr) {
    var js = "javascript:window.NativeBridge = function(name,message){eval(\"messageHandlers.\"+name+\"(message)\")}"

    constructor(context: Context?, attrs: AttributeSet?) : this(context, attrs, 0)
    constructor(context: Context?) : this(context, null)

    init {
        settings.javaScriptEnabled = true
        settings.cacheMode = WebSettings.LOAD_NO_CACHE
        settings.loadWithOverviewMode = true
        settings.useWideViewPort = true
        settings.layoutAlgorithm = WebSettings.LayoutAlgorithm.SINGLE_COLUMN
        addJavascriptInterface(this, "messageHandlers")
        webChromeClient = WebChromeClient()
        webViewClient = object : WebViewClient() {
            override fun onPageFinished(view: WebView?, url: String?) {
                super.onPageFinished(view, url)
                loadUrl(js)
            }
        }
    }
}